### ISSUE TITLE
v1.0.0 Roadmap

### ISSUE THREAD

**Issue by jeff-zucker** [jeff-zucker/solid-file-client#59]
**Before release of v1.0.0**

- COMPLETE HIGH LEVEL METHODS
  - [x] Connection methods: login, popupLogin, currentSession, logout
  - [x] Refactor folderUtils as a class using N3 
  - [x] Refactor processFolder (without getLinks)
  - [x] readFile
  - [x] updateFile
  - [x] shims (pass-through to API) for readHead, deleteFile, 
  - [x] shims for deleteFolder, moveFile, moveFolder
  - [ ] add logging for users
  - [ ] get blob working in node
  - [ ] code cleanup
- COMPLETE API METHODS 
   - [x] createFile (maybe used by copy)
   - [x] createFolder
   - [x] createItem
   - [x] copyFile (used by copyFolder)
   - [x] copyFolder (used by copy, move, rename)
   - [x] and other SolidApi and SolidFileClient methods
- TESTS
  - [x] tests for new methods as we complete them
  - [x] Refactor high-level-crud-test (after test environment complete)
  - [ ] Make sure all tests, including blobs pass with  file:// and app:// and in browser
    - [ ] Re-test with NSS 5.x when it is more stable
- TESTING ENVIRONMENT
  - [x] Add basic logging (fetch and probably a few others in the beginning. The others while developing)
- DOCUMENTATION
  - [ ] low-level methods and option flags
  - [ ] general intro & high-level methods
  - [ ] Change log and migration instructions
  - [ ] Examples for in-browser and node usage
  - [ ] Live examples (e.g. upload)
- FINAL TASKS
    - [ ] Tidy up the codebase - improve code & readability, update JSDocs
    - [ ] Credit collaborators - let me know how you want to be listed in package.json and README
    - [ ] publish to github & npm (cdn happens automatically), announce in forum, chat, twitter
    -  [ ] celebrate! fireworks!

**Maybe before release of v1.0.0**
- [x ] Fix localStorage so app:// can be used in tests

**Future releases**
- [ ] Add Travis CI
- [ ] Enable https testing in jest
- [ ] a readHeaders method that returns a hash of all headers from headers.keys()
- [ ] support getLinks, .acl and .meta, and the withLinks options on various methods
- [ ] Implement OPTIONS in solid-rest
- [ ] create a readAcl method to display user perms based on Otto's package
- [ ] mock login & session methods
- [ ] maybe? re-introduce fetch API response objects rather than throw

**Comment by Otto-AA** [comment]
Here's my view view on the roadmap

# (1) Finish the test environment
- [ ] Fix error for https solid-auth-cli (#58 )
- [ ] Add basic logging (fetch and probably a few others in the beginning. The others while developing)

# (2) Finish methods in SolidApi
Those which will be used by other methods should be implemented first, so the others can already work with a (at least nearly) finished version of e.g. readFolder. This also includes discussing how exactly they should look like and how they behave in certain scenarios. When we write down the results of the discussion in a more or less structured way this could be used for tests and documentation. I'd add the tests at the same time, documentation later on.
- [ ] readFolder (will be used by all recursive methods; includes possible rework of folderUtils)
- [ ] createFile (maybe used by copy)
- [ ] putFile (maybe used by copy)
And with a bit lower priority imo:
- [ ] copyFile (used by copyFolder)
- [ ] copyFolder (used by copy, move, rename)
- [ ] and other SolidApi and SolidFileClient methods

# (3) Tidy up of the code base
Includes making the code more readable, better structured if possible, adding comments to explain unintuitive code (if necessary) and updating JSDocs.

# (4) Adding examples and writing documentation
In this stage I would create several examples for in-browser and node usage as html and js files. Also create a manually written documentation plus the one derived from JSDoc. Also include a changelog and migration guide.

# (5) Releasing v1.0.0
Publishing it to npm, the cdn, as github release and mention the new version on gitter and the forum.

# (6) Celebrate
We've done it! :fireworks:

**Comment by Otto-AA** [comment]
So yeah, we don't have to follow it step by step obviously, but I think it would be a good guideline. The main focus for me now is to fix the error in the test setup with solid-auth-cli+jest. And readFolder also is important right now. These are the two things I'd like to be finished before continuing with the others.

**Comment by jeff-zucker** [comment]
I've updated the roadmap to mix in Otto's suggestions as well as a few random thoughts that questions that occurred to me.

**Comment by jeff-zucker** [comment]
Just remembered to add "test on both NSS 4.x and NSS 5.x"

**Comment by jeff-zucker** [comment]
Oh my goodness, I've discovered how .meta files are meant to work and they are very, very cool.  There are no .meta files for resouces, only for containers.  But in the container's resource you can refer to both the container and its contents.  So if you have this in the .meta of a container:
```
   <> schema:label "Family Photos".
   <foo.png> schema:label "Camping at Yosemite".
```
When you do a GET of the container (not it's .meta, the container itself) you will see this:
```
 <> a ldp:BasicContainer;
     schema:label "Family Photos";
     ldp:contains <foo.png>.

  <foo.png> a png:Resource;
     schema:label "Camping at Yosemite".
```
Which makes tagging photos and music dead simple.  And means all of the complexity of looking for the resource .meta files is unnecessary.

**Comment by Otto-AA** [comment]
Thanks for the merging of the roadmaps, looks good to me :)

> Fix error for https solid-auth-cli (#58 ) <<< are you referring to the jest problem or something in solid-auth-cli?

Getting it working in jest. Highly likely not because of solid-auth-cli, but I don't know what causes this error yet.

Regarding .meta, I've [answered](https://github.com/jeff-zucker/solid-file-client/issues/42#issuecomment-505288616) in the background issue, as I don't think it belongs here.

**Comment by Otto-AA** [comment]
And why did you put blob to future releases? As mentioned before, it would break things if we use .buffer in solid-file-client. We can't accept a random auth client then (and not even solid-auth-client), because they probably don't add a buffer method to the response. Using .blob would be fetch-spec compliant and work with all clients using window.fetch (in particular solid-auth-client). Therefore the use of blob instead of buffer is critical and has to be done before the release.

**Comment by jeff-zucker** [comment]
>
> And why did you put blob to future releases? As mentioned before, it would
> break things if we use .buffer in solid-file-client.
>
Um, solid-file-client has been operating just fine for over a year without
that having broken anything.  A single sanity check of if(typeof buffer
!="undefined") or check to see if the script is running in browser would
prevent it from causing problems to browser users.

> We can't accept a random auth client then (and not even
> solid-auth-client), because they probably don't add a buffer method to the
> response.
>
So we check to see if they're not using solid-auth-cli and throw a "your
client doesn't support blobs" message for them

> Using .blob would be fetch-spec compliant and work with all clients using
window.fetch (in particular solid-auth-client). Therefore the use of blob
instead of buffer is critical and has to be done before the release.

So, if you have a solution, feel free to implement it.

**Comment by Otto-AA** [comment]
> Um, solid-file-client has been operating just fine for over a year without that having broken anything. A single sanity check of if(typeof buffer !="undefined") or check to see if the script is running in browser would prevent it from causing problems to browser users.

I still don't know how the old version works, but I think that either `res.headers._headers` ([here](https://github.com/jeff-zucker/solid-file-client/blob/master/src/index.js#L228)) is undefined and therefore res.buffer() doesn't even get called in the browser, or that it is polyfilled somewhere. I expect the first to be the case. Anyway, if we want to use .buffer in the browser we have to use a polyfill (which we currently don't and which seems unnecessary imo) or require the client to manually add .buffer (which is definitely not what I want, because it would break the use together with solid-auth-client).

I've also tried to use `fileClient.readFile('https://httpbin.org/image/png')` with our current version and it has thrown an error that .buffer is undefined at [this line](https://github.com/jeff-zucker/solid-file-client/blob/otto-changes/src/SolidFileClient.js#L137). So I am 99% sure that it breaks the use with solid-auth-client and other users in the browser.

The solution I see is to simply replace all .buffer() calls with .blob(). I think it is pretty much the same, so could you try that in your readFile method? If I replace it it doesn't throw and resolves with a Blob, which seems like the intended behaviour.

**Comment by jeff-zucker** [comment]
I never claimed it worked in the browser.  There's nothing in the
documentation that says it does.  It has worked fine in nodejs and I've
used it to upload music and photos.  If replacing it with blob() is all it
takes, well then of course, let's do that.  I'll try.

**Comment by Otto-AA** [comment]
Would you accept a PR for solid-rest which makes the response like the standard window.fetch response (including .buffer())? I would try to copy the relevant files from node-fetch and then modify some parts if necessary. Having the same Response object as a standard fetch (in particular .blob) would make developing of copy/move faster, therefore I'd like to have that before finishing copyFile.

**Comment by Otto-AA** [comment]
Or I think something like this would also work:
```javascript
const { Response } = require('node-fetch')
new Response('some data', { status: 401, ...options})
```

**Comment by jeff-zucker** [comment]
Yes sure.  I have not been resistant  to doing that because it's a bad
idea, but because I just need to limit how many things I deal with at once
- too many balls in the air.  I'd much appreciate a PR.

**Comment by Otto-AA** [comment]
Ok fine. Just wanted to ask before doing unnecessary work :)
I'll give it a try this weekend.

**Comment by jeff-zucker** [comment]
I saw your forum post about reading .acls did you see it on the roadmap above?

**Comment by jeff-zucker** [comment]
Given rdf-query (now called rdf-easy), it would be pretty simple to read the .acls

**Comment by Otto-AA** [comment]
> I saw your forum post about reading .acls did you see it on the roadmap above?

I mean parsing an acl file as well as creating new permissions. I've seen it in the roadmap, but I think it would be useful if there's already a library for that, and if not I would do it outside of solid-file-client and then import that module (because I suspect it to be useful without solid-file-client too). That's why I've asked there.

> Given rdf-query (now called rdf-easy), it would be pretty simple to read the .acls

Yeah, I would use N3 or some wrapper for parsing and writing. I haven't taken a look at rdf-query yet, so I don't know what I'd pick. But it is still just an idea, no concrete plan for now :)

**Comment by jeff-zucker** [comment]
Cool. Let me know if I can be of help.

**Comment by bourgeoa** [comment]
I made a PR to **solid-rest** to solve recursive PUT in `file://`
All tests passes for `app:` and `file:`

**Comment by jeff-zucker** [comment]
May I ask why you went and added semi-colon  all over  my code?

**Comment by bourgeoa** [comment]
I forgot do delete them. It is not in purpose I added them when I tried the
understand the code.
Sorry as I told I never learn any basics in coding or style.

Le ven. 5 juil. 2019 à 18:20, Jeff Zucker <notifications@github.com> a
écrit :

**Comment by jeff-zucker** [comment]
I wondered if maybe you knew of a reason they were better.  I've found that I have a high percentage of typos adding semi-colons and I've gotten used to the clean look of the code without them.  But  AFAIK, it's just a matter of personal taste.  Anyway, not a big problem, thanks for the PR.

**Comment by Otto-AA** [comment]
Semicolons aren't really needed in JS, you can use them, but some actually advice not to do this. For instance the standard style linter I've added to solid-file-client enforces a “never use semicolons” rule because they are not necessary (you can read up here on why it does if you want: https://feross.org/never-use-semicolons/). But after all, it is mostly a design question

**Comment by Otto-AA** [comment]
With #69 all the composed methods I've written (copyFile, copyFolder, move, rename, deleteFolderRecursively) *should* work now. The tests in app:ls work. The ones for file:// fail because it doesn't support node-fetch blobs (see below*). It is not tested in the browser, but I wouldn't expect much trouble there.

I think the three options they accept (overwriteFiles, overwriteFolders and createPath) are good the way they are now. The other options (copyAcl, copyMeta) are ignored for now and can be implemented/removed when this discussion has cleared up.

### Documentation 

I think for the documentation it would be cool to have one or more live examples in html. e.g. a little "Solid Upload" app, which has an file input and lets you choose a location and it uploads it. If we have some examples which cover the basic usages it would be make getting started much easier imo.

Apart from that it should contain a getting started guide, an overview about all high-level methods and some automatically generated JSDoc documentation. As mentioned before, I think it would make sense to use docsify for better structuring (I wouldn't put everything into the same document. And the sidebar of docsify is good for having multiple docs and sections).


### Testing

The test environment is ok imo. It would be cool if we fix the error with jest and the https:// tests, but I guess that won't happen soon. I will change the contextSetup.js script so it doesn't require login. So as @bourgeoa suggested we can bypass this error by creating a folder where everyone has read/write access and then use it without logging in. I will also try to add a little "delay" mechanism which makes sure we don't send 100 requests at once, but caps it at a rate per second.

> Make sure everything works in NSS 5.x

Afaik the new inrupt [pod-server](https://github.com/inrupt/pod-server) is nearly ready. If we could test there too instead of only NSS that would be awesome.

Also it seems to make sense to add travis CI, so we know if a commit/pr breaks some automated solid-rest tests.

### Blobs

*regarding the blobs: I would suggest to always expect and use blobs in the responses. This behaviour is supported by all browsers (ie it works with solid-auth-client) and should be possible to support in nodejs. It has a stream and arrayBuffer method, so that seems doable in general.

For solid-rest, I think the only thing to change to add blob support to file:// is to check [here](https://github.com/jeff-zucker/solid-rest/blob/master/src/file.js#L12) whether or not the passed parameter has a ".stream()" method and if yes return that instead. I believe this is enough to make it work in file://

**Comment by Otto-AA** [comment]
Is this detailed enough, or have I forgotten something?

**Comment by jeff-zucker** [comment]
Looks  great.  I need to actually experiment with blobs because from your description I am not really clear, but let me try and get back to you with questions.

**Comment by Otto-AA** [comment]
> Looks great. I need to actually experiment with blobs because from your description I am not really clear, but let me try and get back to you with questions.

Just in case it isn't clear: Nodejs has no Blob support inbuilt, they are just imitated by node-fetch. But it should work pretty similar to the ones in the browser.

**Comment by jeff-zucker** [comment]
I have edited the roadmap, adding in Otto's last comments, checking off things I consider done, adding a few more to-dos, and moving some things from before release to after release and some things (blobs) in the other direction.  If you disagree with any of my changes or have more, please speak up.

If we all agree that with the roadmap, my guess is that we could be ready for a release in early August.   Otto, if you don't like the idea of it being released before you have time to fully review, we could consider holding off until September.

**Comment by Otto-AA** [comment]
> Add basic logging (fetch and probably a few others in the beginning. The others while developing)

I think this would be good to make a bit more detailed and to provide an option to the user to enable it in their app. I'd suggest following option in the constructor:
```javascript
new SolidFileClient(fetch, { enableLogging: true }) // Calls debug.enable('solid-file-client:*')
// and for more granular control also allow a string
new SolidFileClient(fetch, { enableLogging: 'solid-file-client:fetch*' }) // Calls debug.enable('solid-file-client:fetch*')
```
https://www.npmjs.com/package/debug#set-dynamically

> Make sure all tests pass with file:// and app:// and in browser

This is not really true for the copyFile/copyFolder/deleteFolderRecursively/move/rename methods as they haven't been tested properly in the browser. Also file:// doesn't work yet for some of them due to the blob issue.

Apart from those I agree with the roadmap :+1:

**Comment by Otto-AA** [comment]
> If we all agree that with the roadmap, my guess is that we could be ready for a release in early August. Otto, if you don't like the idea of it being released before you have time to fully review, we could consider holding off until September.

I will be home in the end of July for a few days, so I can quickly look through the changes and if I think something should be changed/done before the release I can make a remark then. So beginning of August would be fine to me.

**Comment by Otto-AA** [comment]
I've quickly tried out the blob.stream() method for file:// and it works flawlessly from what I've seen. All tests of solid-rest pass and also all SolidApi tests pass with file:// when this blob check is added. But if you want to test it out properly before releasing, there's no need to hurry from my side.

**Comment by Otto-AA** [comment]
By the way, there is some issue with the test setup when test-folder already exists. Manually deleting it before running file:// tests temporarily solves it

**Comment by Otto-AA** [comment]
Apparently, the isomorphic fetch library used in solid-auth-cli uses [whatwg-fetch](https://www.npmjs.com/package/whatwg-fetch) when making http requests from node. And in this case it doesn't respond with a blob method, so copyFile and all which rely on it fail. Not sure yet, how we should handle this best

**Comment by Otto-AA** [comment]
So the blob summary:
Browser:
- https: *should work*
- app://ls: *not yet*
- app://memory: *not yet implemented, but would work*
- solid-auth-client: *should work*

node:
- node-fetch: works with file system, *should work* for https
- app://ls: works as it falls back to in memory
- app://fs: works with the PR I've made
- solid-auth-cli: https://: doesn't work because .blob somehow isn't defined

As far as I know, if we do https requests from node, we should also use node-fetch (which would work with our blob implementation). whatwg-fetch states that it is browser-only, so it wouldn't make sense if it is called. I will take another look to see which library is called here...

**Comment by Otto-AA** [comment]
Ok, after some digging into the the node_modules dependencies, I've found out that ismorphic-fetch is using an outdated version. It specifies `"node-fetch": "^1.0.1"` (=1.x) in its dependencies, so it sticks with the newest 1.x version of node-fetch. But node-fetch is already at v2.0.6, the blob support was added in v2.0.0 as experimental feature (see their [changelog](https://github.com/bitinn/node-fetch/blob/master/CHANGELOG.md#v200)).

TL;DR: isomorphic-fetch uses an outdated version of node-fetch which means it doesn't support blobs

My suggestion for solid-auth-cli is, to use [cross-fetch](https://github.com/lquixada/cross-fetch) instead. It is basically the same as isomorphic-fetch, but with newer versions. And the last commit there was last week, whereas the last commit in isomorphic-fetch was in 2016.

I don't think we need to change anything in solid-file-client because of this. I think it is enough if it works with node-fetch and cross-fetch, isomorphic-fetch is too outdated to support imho.

**Comment by Otto-AA** [comment]
So... what is the current status of this project?

Here are the things which are still missing afaik. Are there other things too?

- add documentation (describe methods, quickstart, examples)
- finish all TBD/TODOs in the code base
- consider moving rdf-query.js to another package?

**Comment by Otto-AA** [comment]
> My suggestion for solid-auth-cli is, to use cross-fetch instead. It is basically the same as isomorphic-fetch, but with newer versions. And the last commit there was last week, whereas the last commit in isomorphic-fetch was in 2016.

And this is still something I think is necessary for the blob support. If you want I can try to make this change in solid-auth-cli, but I as I'm not familiar with this project it is probably easier if you try it.

**Comment by jeff-zucker** [comment]
Current status - I'm disgusted with the lack of progress on NSS 5.x and am
unsure how to proceed given the news that it will no longer be maintained.
I am working on the documentation and will get back to the other tasks this
coming week.

I'm unclear about why you suggest moving rdf-query.js.  It is basically
just a way to  talk to N3 from solid-file-client and not particularly
useful on its own.   I have released a separate package rdf-easy which is
based on rdflib and not particularly useful for solid-file-client, so the
two are unrelated.

**Comment by jeff-zucker** [comment]
I will take a stab at blob support but it will be in solid-rest-file, not in solid-auth-cli.  Solid-auth-cli is an exact duplicate of solid-auth-client and neither package is responsible for reading or writing to disk, only sending requests to do so that are responded to  by the server.  In this case a request for file:/// does not go to a server, it goes to solid-rest-file which is responsible for the actual reading and writing.  Am I missing some reason that is not where blob should be?

**Comment by Otto-AA** [comment]
> Current status - I'm disgusted with the lack of progress on NSS 5.x and am
unsure how to proceed given the news that it will no longer be maintained.

Oh, I didn't know that. Are they completely dropping maintenance for it?

> I am working on the documentation and will get back to the other tasks this
coming week.

What is the structure you have in mind for the documentation?

> I'm unclear about why you suggest moving rdf-query.js.  It is basically
just a way to  talk to N3 from solid-file-client and not particularly
useful on its own.   I have released a separate package rdf-easy which is
based on rdflib and not particularly useful for solid-file-client, so the
two are unrelated.

Ok fine

**Comment by Otto-AA** [comment]
> I will take a stab at blob support but it will be in solid-rest-file, not in solid-auth-cli. Solid-auth-cli is an exact duplicate of solid-auth-client and neither package is responsible for reading or writing to disk, only sending requests to do so that are responded to by the server. In this case a request for file:/// does not go to a server, it goes to solid-rest-file which is responsible for the actual reading and writing. Am I missing some reason that is not where blob should be?

The problem I see is the use of the isomorphic-fetch package in solid-auth-cli, because it is unmaintained since 3.5 years and therefore uses an outdated version of node-fetch which doesn't support blobs. Afaik the cross-fetch package does the same thing and is still being maintained, so I'd suggest to switch to it. So I don't see anything you could do in solid-rest-file for the blobs.
(see my [earlier comment](https://github.com/jeff-zucker/solid-file-client/issues/59#issuecomment-510402051) for a bit more details)

**Comment by jeff-zucker** [comment]
I will look at solid-auth-cli's isomorphic fetch but that has nothing to do
with reading files on a server.  It is the fetch that I (and at the time
solid-auth-client) use to perform https:// fetches.  Solid-auth-cli does
not do any file:// fetches,  it hands thoseoff to solid-rest-file.fetch and
so are never touched by isomorphic fetch.

**Comment by jeff-zucker** [comment]
Above I said " has nothing to do with reading files on a server", when I
meant "from a file system".

**Comment by jeff-zucker** [comment]
>
> Current status - I'm disgusted with the lack of progress on NSS 5.x and am
> unsure how to proceed given the news that it will no longer be maintained.
>
> Oh, I didn't know that. Are they completely dropping maintenance for it?
>
I'll ask that question directly to the NSS devs.

**Comment by Otto-AA** [comment]
The problem for blobs is not when using the file system. It is when making https requests when using solid-auth-cli in node. So node->web server is not working. And for this case solid-auth-cli uses the isomorphic-fetch package for making the request (which in turn uses an outdated version of node-fetch and therefore the response doesn't has a .blob() method).

**Comment by Otto-AA** [comment]
You can take a look at my comment above for a table of what should work with blobs and what doesn't: https://github.com/jeff-zucker/solid-file-client/issues/59#issuecomment-510389084

**Comment by jeff-zucker** [comment]
OK, thanks, I'll look at it later this week, sorry to be so slow to
understand.

**Comment by Otto-AA** [comment]
May I ask what the plans for the release of v1.0.0 currently are? I haven't been that active in the Solid project recently (and likely won't be in the near future), but I guess that we have stable pod versions to test by now. Or is this still an issue?

**Comment by jeff-zucker** [comment]
Apologies, I have not been able to get back full time to Solid so far but hope I will be able to soon.  I'd like to release 1.0.0 in a week or so.  AFAIK, the NSS issues are all solved, but I haven't done full testing.

**Comment by Otto-AA** [comment]
What are the tasks we still want to do before the v1.0.0 release?

- Finishing SolidFileClient
- - add trackSession
- - finish high level methods
- - Agreeing on Error Interface
- Documentation
- - documentation format (plain markdown? docsify? ...?)
- - High level documentation
- - Examples (short code excerpts and websites like Solid Copy/Upload)
- Tyding up
- - Removing unused comments
- - Linting with standardjs
- - Checking metadata (package.json, README, links, ...)
- - Checking dependencies

From my point of view is SolidApi pretty much finished. I'll take a look at the withAcl part when I have more free time (likely end of next week). The error interface probably still changes.

Would you agree with these? Is something missing?

**Comment by jeff-zucker** [comment]
I have not kept up with changes in solid-auth-client so I'm unclear what trackSession would do and whether newer developments in solid-auth-client need to be reflected in SolidFileClient.  Which leads me to conclude that @Otto-AA  was right when he originally suggested we let users do login using the auth client, and NOT provide pass-throughs in SolidFileClient.  In other words remove all the login and session methods. Thoughts?

**Comment by jeff-zucker** [comment]
I like docsify for local use.  But how can the resulting docs be presented on github or usable by a user who clones or npm installs the package?

**Comment by Otto-AA** [comment]
> In other words remove all the login and session methods. Thoughts?

I'd prefer it that way. That the user is logging in via solid-auth-client (or another library) and not via SolidFileClient. But we should include it in the examples imo.

> trackSession

Calls a callback everytime the session status changes (ie on login and logout). Pretty useful for keeping track of the login status.

**Comment by Otto-AA** [comment]
> I like docsify for local use. But how can the resulting docs be presented on github or usable by a user who clones or npm installs the package?

We could deploy it on github pages: https://docsify.js.org/#/deploy?id=github-pages
And as it is based on markdown files, github will also display them properly. And when the user downloaded them they can also view them locally.

I've quickly tried it out with the solid-acl-parser, you can take a look here: https://otto-aa.github.io/solid-acl-parser/#/

**Comment by jeff-zucker** [comment]
@bourgeoa do you agree with dropping the login/session methods from solid-file-client and requiring users to get the methods directly from the auth packages?  I guess this would be a pretty major change, perhaps we could provide a light wrapper they can optionally use if they don't want to change the code to use the auth methods rather than the solid-fc methods.

**Comment by jeff-zucker** [comment]
Great, thanks, I'll look at those.  I do like docsify, just not sure how
much more complexity I want to add to the build process.

**Comment by bourgeoa** [comment]
Sorry Jeff I don't understand the implications

> dropping the login/session methods from solid-file-client and requiring users to get the methods directly from the auth packages?

**Comment by jeff-zucker** [comment]
> Sorry Jeff I don't understand the implications

Currently there is a SolidFileClient.login() method which simply calls SolidAuthCli.login() and similarly for all the login and session methods.  I propose that we do not include these methods at all.  That users do something like this instead;
```javascript
const auth = require('solid-auth-cli')
const FC = require('solid-file-client')
const fc = new FC(...)
...
await auth.login()  // instead of await fc.login()
fc.readFile(...)
```

**Comment by bourgeoa** [comment]
> Currently there is a SolidFileClient.login() method which simply calls SolidAuthCli.login() and similarly for all the login and session methods. I propose that we do not include these methods at all. 
> 
Why not It is a change and It should be clearly documented.
Anyway we have a lot of changes.

**Comment by bourgeoa** [comment]
For documentation I understand we will use JSdoc.
This is nice for API but lambda user like I, find it very difficult to understand to use. It is a reference for later use.
I liked very much solid-file-client README.md can we have the same also ?

**Comment by Otto-AA** [comment]
Yes, afaik Jeff is already working on some documentation.

**Comment by jeff-zucker** [comment]
I also really dislike JSdoc.  It is OK with me if we include it, but I
intend to have a fully documented README like the one we had before as the
main documentation.  I will be doing it in straight markdown, not docsify
for this version - there are just far too many moving parts.

**Comment by bourgeoa** [comment]
Is there a way for me to help more then review

**Comment by jeff-zucker** [comment]
On Tue, Dec 10, 2019 at 10:03 AM Alain Bourgeois <notifications@github.com>
wrote:

> Is there a way for me to help more then review
>
Some ideas, in no particular order:

Start a document on backwards compatibility.  I can polish the language
later, but it would help if you get down some basic ideas.

Make a github repo of your new solid-ide?

Would you have any interest in preparing some jest tests for
SolidFileClient?  Essentially re-writing crud-node for jest.  I think it
should be done to automate the tests but my knowledge of jest and of
@Otto_A_A's test framework is pretty weak.  I'd also understand if you want
to pass on this.

If none of those appeal and Otto doesn't have others pertaining directly to
solid-file-client, you could take a look at how localStorage works in the
browser (a solid-rest issue, only tangentially related).

**Comment by bourgeoa** [comment]
@jeff-zucker 
I made a github repo [https://github.com/bourgeoa/solid-ide](url)
working with solid-file-client v1.0.0
[https://bourgeoa.github.io/solid-ide/](url)

**Comment by Otto-AA** [comment]
It feels like we are getting close to the final version of v1.0.0 now. If possible I would like to finish it before the 23th December, afterwards I'll likely won't contribute to significant changes and discussions of it. It was a good experience to work with you so far, but it's also time consuming and I should limit that soon. If we don't make it until the 23th, I am sorry that you will have to finish it without me, but I think we are on a good way to be done by then :)

**Comment by jeff-zucker** [comment]
Yes, I agree we are close.  I will try my best to get it done by the 20th.
 I agree that it has been a great experience working with the two of you.
I totally understand that you have other priorities and won't expect you to
be working with us though I hope if we get stuck on something you might
from time to time take a peek.

**Comment by jeff-zucker** [comment]
Excellent!  Thanks!

**Comment by Otto-AA** [comment]
Okidoki, as said earlier I will work less on this project now. If you have any question or want help with something feel free to ask (with a @mention). If it's before Sylvester chances are rather high that you will get an answer. Apart from that I don't think that I'll be active here. I wish you a merry christmas :)

Things I noticed that still should be done:
- TODOs in comments
- Testing with NSS*
- Update of examples (e.g. links and merge options for the Solid Copy example)
- Finish documentation (I don't know the status of it tbh)

*regarding tests with NSS: You can also try to use the http:// prefix with jest. I usually don't use it because it makes lots of http requests, but if you host your own server it should be fine. Here's how to do it:
1. create a folder which everyone can modify (e.g. /private/test-folder/)
2. optionally add .only to the tests you want to run (e.g. describe.only('copyFolder')). If not it will run the whole file
3. If you want to test SolidApi.links.test.js some things could get broken. Before doing that, modify what acl content it will write in SolidApi.links.test.js and TestFolderGenerator.
4. optionally set `httpRequestsPerSecond` in contextSetup (high number if it is your server and you don't care about requests. low if you don't want to spam a server)
5. run the tests: `DEBUG=solid-file-client:* TEST_PREFIX=https:// TEST_BASE_URL=https://your-pod.solid.community/private/test-folder/ npx jest tests/SolidApi.composed.test.js` (replace SolidApi.composed.test.js with the test file you want to test)
6. get popcorn and watch the logs until it is finished...

**Comment by jeff-zucker** [comment]
Thanks a million Otto.  It has been a privilege and a pleasure to work with you.  I'll try to keep questions to a minimum.  A very merry Christmas to you and I hope your coming year is fruitful and mostly Solid-free.

**Comment by jeff-zucker** [comment]
@bourgeoa may I leave the HTTP testing to you?  You have the server and also Solid-IDE.  I will dig into the file:// tests and try to get them working as well as looking at the comments in the code.  I think the documentation is very close to being done, could you look it over and make sure that, especially with the merge options it documents the way it behaves.  I guess I will aim to release on the first of the year.

**Comment by bourgeoa** [comment]
I finally begun the HTTPS tests. I had a SSL let's encrypt certificate incomplete on Node.js on my server (need of ca.cert, not needed in browser)

@Otto-AA 
The only error in SolidApi.core.test.js is in HEAD for folder : 
contentType returns application/octet-stream and not text/turtle.

**Comment by jeff-zucker** [comment]
The moving and copying of links was broken for folder links.  Copying /foo/bar to /foo/bop/ would put the /foo/bar/.acl and /foo/barr/.meta in /foo/ rather than in /foo/bop.  Changes for this in SolidApi.js are marked with "// JZ".

**Comment by jeff-zucker** [comment]
I bumped the version of solid-rest to 1.1.1, make sure to **run npm install!**

**Comment by Otto-AA** [comment]
> The only error in SolidApi.core.test.js is in HEAD for folder :
contentType returns application/octet-stream and not text/turtle.

I took a look at it again, and its an old NSS bug: https://github.com/solid/node-solid-server/issues/454

**Comment by jeff-zucker** [comment]
Wow, what a serious and long-standing bug!  Does this impact our internals?
In other words, do we use HEAD to find content-type or other info in our
code?

**Comment by bourgeoa** [comment]
I checked the code and we do use HEAD for itemExists, createFolder and getItemLinks.
In none of these cases do we use the contentType returned from HEAD
Everywhere contentType is returned from GET.

I don't see this NSS bug to have any impact on solid-file-client.

**Comment by jeff-zucker** [comment]
> check dependencies (I'm never sure where to put them. I'd guess that all peer dependencies should be put to devDependencies

@bourgeoa, this makes sense to me to move the peerDependencies to devDependencies.  Do you agree?

**Comment by bourgeoa** [comment]
I have no experience.
- peer dependencies seems to be used only for publishing : 
```
Peer dependencies are a special type of dependency that would only ever come up if you were publishing your own package.

Having a peer dependency means that your package needs a dependency that is the same exact dependency as the person installing your package. This is useful for packages like react that need to have a single copy of react-dom that is also used by the person installing it.
```

- npm install uses all dependencies
- It seems that dev-dependencies shall only include : (Here are the responses found by googling) 
```
devDependencies should contain packages which are used during development or which are used to build your bundle, for example, mocha, jscs, grunt-contrib-watch, gulp-jade and etc. These packages are necessary only while you are developing your project, also ESlint is used to check everything during building your bundle.
```
```
Some examples of packages that would only be required during development are babel plugins and presets, test runners and linter packages.
```

**Comment by Otto-AA** [comment]
My view on the packages is: If we bundle everything into dist and publish that, then the users don't need any other packages installed to use our code. Hence everything we have would be devDependencies.

**Comment by jeff-zucker** [comment]
Solid-auth-cli and Solid-rest will never be in a bundle though, right?  But
having them in devDependencies means that they will be available for
running tests and command-line use within the installation folder.

**Comment by Otto-AA** [comment]
> Solid-auth-cli and Solid-rest will never be in a bundle though, right?  But
having them in devDependencies means that they will be available for
running tests and command-line use within the installation folder.

Yes.

**Comment by jeff-zucker** [comment]
Version 1.0.0. is now released, closing  background issues.
